### PR TITLE
Add batchUpdate() to QueryBuilder, Command and Migration

### PR DESCRIPTION
**Pull request by WarLikeLaux**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Tests added?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | https://github.com/yiisoft/yii2/discussions/20759

## What does this PR do?

Adds `batchUpdate()` to `QueryBuilder`, `Command`, and `Migration` for updating multiple rows with per-row column values in a single query via `CASE WHEN ... THEN ... END`.

```php
$db->createCommand()->batchUpdate('user', [
    ['id' => 1, 'name' => 'Tom', 'status' => 1],
    ['id' => 2, 'address' => 'new address'],
], 'id')->execute();
```

Generates:

```sql
UPDATE user SET
  name=CASE WHEN id=1 THEN 'Tom' ELSE name END,
  status=CASE WHEN id=1 THEN 1 ELSE status END,
  address=CASE WHEN id=2 THEN 'new address' ELSE address END
WHERE id IN (1, 2)
```

Each row may contain different columns. Missing columns are left unchanged. Rows with only the key column are skipped.

MSSQL and PostgreSQL overrides handle `normalizeTableRowData` (same pattern as existing `update()`).

### Tests

**QueryBuilderTest.php (11 tests):**
- SQL generation: sparse rows, Expression values, null keys, only-null keys
- Validation: missing key, duplicate keys, invalid row type, invalid key value (Expression)
- Schema: with/without table schema, Traversable row

**CommandTest.php (4 tests):**
- Integration: insert 4 rows, batchUpdate with sparse columns, verify result
- Edge cases: empty updatable columns returns 0, empty ArrayObject rows, SQL with Expression params

### CODE REVIEW

**Comment by WarLikeLaux** [comment]
Opening as draft. This is my first feature PR to Yii2 - previous contributions were test coverage only. Would appreciate any feedback on the API design (method signature, naming, SQL generation approach) before marking as ready for review.

**Comment by terabytesoftw** [comment]
@rob006 You could take a look, thanks.

**Comment by Tigrov** [comment]
It's a good idea, but in my opinion it should be implemented as `insertBatch()` (`batchInsert()` in Yii2)
https://github.com/yiisoft/db/blob/dffa745e646d475b6dd993a421c26a088e1c4f0f/src/QueryBuilder/AbstractDMLQueryBuilder.php#L67-L89

with adding `ON CONFLICT (...) DO UPDATE SET ...` as in `upsert()` (for each DBMS it can be different)
https://github.com/yiisoft/db-pgsql/blob/01efbd3211d41d0575912af13a0728cd8c7ccd84/src/DMLQueryBuilder.php#L71-L87

### Signature of the method
```php
interface QueryBuilderInteface
{
   // ...
   public function upsertBatch(
        string $table,
        iterable $insertRows,
        array $insertColumns = [], // not sure if we can remove it
        array|bool $updateColumns = true,
        array &$params = [],
    ): string
   // ...
}
```

`$updateColumns` should contain list of `column => value|expression` to update or `bool` similar to `upsert()`
also `column` can contain `Case` expression (optionally).

### How to implement
A simple implementation example: copy [`upsert()`](https://github.com/yiisoft/db-pgsql/blob/01efbd3211d41d0575912af13a0728cd8c7ccd84/src/DMLQueryBuilder.php#L63-L88) method and replace
```php
$insertSql = $this->insert($table, $insertColumns, $params);
```

with

```php
$insertSql = $this->insertBatch($table, $insertRows, $insertColumns, $params);
```

**Comment by rob006** [comment]
@Tigrov  I'm using custom implementation of `upsertMultiple()` for this purpose in my project (MySQL) and this approach has several weird quirks. 
1. In query you need to specify all required columns, even the ones you don't want to touch.
2. There is always a risk that you insert new record, if specified ID doesn't exist (record could be delated in the meantime by concurrent process).
3. When I had more than one unique column in table, MySQL reserved IDs for all specified records, even if no records were inserted in the end. This created huge gaps in primary key sequence. I had to remove unique constraint on `slug` column because of that.

All this could be avoided if we use `SET column=CASE...` approach.

File: framework/db/mssql/QueryBuilder.php [hunk 1]
```diff
@@ -619,6 +620,86 @@ public function update($table, $columns, $condition, &$params)
...
+                }
+                if (count($row) !== $columnCount) {
+                    throw new InvalidArgumentException(
+                        'Each batch update row must have exactly ' . $columnCount . ' values when $columns is specified.'
+                    );
+                }
+                $resolvedRows[] = array_combine($columns, array_values($row));
+            }
+            $rows = $resolvedRows;
+            $columns = [];
```
Comment (Tigrov) [inline framework/db/mssql/QueryBuilder.php]:
```suggestion
```
There is no need to reset `$columns`

File: framework/db/mssql/QueryBuilder.php [hunk 2]
```diff
@@ -619,6 +620,86 @@ public function update($table, $columns, $condition, &$params)
...
+        }
```
Comment (Tigrov) [inline framework/db/mssql/QueryBuilder.php]: Can it be moved to separate method, if this part is the same for each DBMS?

File: framework/db/mssql/QueryBuilder.php [hunk 3]
```diff
@@ -619,6 +620,86 @@ public function update($table, $columns, $condition, &$params)
...
+        if (empty($keys)) {
+            $tableSchema = $this->db->getSchema()->getTableSchema($table);
+            if ($tableSchema !== null && !empty($tableSchema->primaryKey)) {
+                $keys = $tableSchema->primaryKey;
+            } else {
+                throw new InvalidConfigException(
+                    'The $keys parameter must be specified because the table "' . $table . '" has no primary key defined.'
+                );
+            }
+        }
```
Comment (Tigrov) [inline framework/db/mssql/QueryBuilder.php]: Can it be moved to separate method?

File: framework/db/mssql/QueryBuilder.php [hunk 4]
```diff
@@ -619,6 +620,86 @@ public function update($table, $columns, $condition, &$params)
...
+            $name = ':' . $matches[1];
+            if (!array_key_exists($name, $params)) {
+                return $matches[0];
+            }
+
+            $expandedName = ':qp' . count($expandedParams);
+            $expandedParams[$expandedName] = $params[$name];
+
+            return $expandedName;
+        }, $sql);
```
Comment (Tigrov) [inline framework/db/mssql/QueryBuilder.php]: Is it correct?
`$params` was passed by the link in `parent::batchUpdate()` and has the same placeholders as generated inside the parent method. Why need replace them?

File: framework/db/mssql/QueryBuilder.php [hunk 5]
```diff
@@ -619,6 +620,86 @@ public function update($table, $columns, $condition, &$params)
...
+                    $keyValues[$keyCol] = $row[$keyCol];
+                    unset($row[$keyCol]);
+                }
+            }
+            $row = $this->normalizeTableRowData($table, $row, $params);
+            foreach ($keyValues as $keyCol => $keyValue) {
+                $row[$keyCol] = $keyValue;
+            }
+            $normalizedRows[] = $row;
+        }
```
Comment (Tigrov) [inline framework/db/mssql/QueryBuilder.php]: Perhaps some parts of the same code also can be move to separate method.

File: framework/db/pgsql/QueryBuilder.php [hunk 6]
```diff
@@ -457,6 +458,67 @@ public function update($table, $columns, $condition, &$params)
...
+                );
+            }
+        }
+
+        $normalizedRows = [];
+        foreach ($rows as $row) {
+            if (!is_array($row)) {
+                $normalizedRows[] = $row;
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (WarLikeLaux) [reply framework/db/pgsql/QueryBuilder.php]: Good catch. Fixed - added `iterator_to_array()` before the `is_array()` check in both mssql and pgsql normalization loops.

File: framework/db/mssql/QueryBuilder.php [hunk 7]
```diff
@@ -619,6 +620,86 @@ public function update($table, $columns, $condition, &$params)
...
+                );
+            }
+        }
+
+        $normalizedRows = [];
+        foreach ($rows as $row) {
+            if (!is_array($row)) {
+                $normalizedRows[] = $row;
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (WarLikeLaux) [reply framework/db/mssql/QueryBuilder.php]: Good catch. Fixed - added `iterator_to_array()` before the `is_array()` check in both mssql and pgsql normalization loops.

File: framework/db/mssql/QueryBuilder.php (hunk 1 above)
Comment (WarLikeLaux) [reply framework/db/mssql/QueryBuilder.php]: Removed.

File: framework/db/mssql/QueryBuilder.php (hunk 2 above)
Comment (WarLikeLaux) [reply framework/db/mssql/QueryBuilder.php]: Extracted `resolveColumnNames()` and `resolveKeys()` as protected methods in the base `QueryBuilder`. All four overrides (base, mssql, pgsql, oci) now call them.

File: framework/db/mssql/QueryBuilder.php (hunk 3 above)
Comment (WarLikeLaux) [reply framework/db/mssql/QueryBuilder.php]: Covered by the same change as above, `resolveKeys()`.

File: framework/db/mssql/QueryBuilder.php (hunk 4 above)
Comment (WarLikeLaux) [reply framework/db/mssql/QueryBuilder.php]: Yes, it's needed. The parent's `batchUpdate()` reuses the same key placeholder in both `CASE WHEN id=:qp0` and `WHERE id IN (:qp0)`. MSSQL's sqlsrv PDO driver doesn't support using a named parameter more than once in a statement, so this block renumbers every occurrence to make each one unique. Added a comment explaining this.

File: framework/db/mssql/QueryBuilder.php (hunk 5 above)
Comment (WarLikeLaux) [reply framework/db/mssql/QueryBuilder.php]: The "extract keys -> normalize -> restore keys" loop is the same in mssql and pgsql, but `normalizeTableRowData()` has different signatures (`pgsql` without `&$params`, `mssql` with `&$params`), so a common method would mean changing the private methods too. Left as is for now, only ~15 lines of duplication.

File: framework/db/mssql/QueryBuilder.php (hunk 4 above)
Comment (Tigrov) [reply framework/db/mssql/QueryBuilder.php]: Aha, got it. Is it possible to generate `where` placeholders using `keys` placeholders inside parent method to avoid this part of code?

Comment (WarLikeLaux) [reply framework/db/mssql/QueryBuilder.php]: Good point. Refactored the parent's `batchUpdate()` to generate unique placeholders for every occurrence - each CASE/WHEN clause and the WHERE clause now bind their own params. The MSSQL renumbering block is removed entirely.

File: framework/db/mssql/QueryBuilder.php (hunk 5 above)
Comment (WarLikeLaux) [reply framework/db/mssql/QueryBuilder.php]: Reconsidered this. You're right, for a framework it's worth extracting properly. Moved `normalizeTableRowData()` to the base `QueryBuilder` as a protected no-op (overridden in mssql and pgsql), and the shared "extract keys -> normalize -> restore keys" loop is now `normalizeBatchUpdateRows()` in the base class. Both overrides just call it.

File: framework/db/QueryBuilder.php [hunk 8]
```diff
@@ -501,6 +502,304 @@ public function batchInsert($table, $columns, $rows, &$params = [])
...
+            $keyHashParts = [];
+            foreach ($keys as $keyCol) {
+                if (!array_key_exists($keyCol, $row)) {
+                    throw new InvalidArgumentException(
+                        'Each batch update row must contain the "' . $keyCol . '" column.'
+                    );
+                }
+
+                $kv = isset($columnSchemas[$keyCol]) ? $columnSchemas[$keyCol]->dbTypecast($row[$keyCol]) : $row[$keyCol];
+                if ($kv instanceof ExpressionInterface || (!is_scalar($kv) && $kv !== null)) {
```
Comment (Tigrov) [inline framework/db/QueryBuilder.php:567]: if `$kv instanceof ExpressionInterface` can it be built using `$this->buildExpression($kv, $params)`?

File: framework/db/QueryBuilder.php [hunk 9]
```diff
@@ -501,6 +502,304 @@ public function batchInsert($table, $columns, $rows, &$params = [])
...
+                } else {
+                    $valueSql = $this->bindParam($value, $params);
+                }
+                $casesByColumn[$column][] = 'WHEN ' . implode(' AND ', $whenParts) . ' THEN ' . $valueSql;
+            }
+        }
+
+        if (empty($casesByColumn)) {
+            return '';
+        }
```
Comment (Tigrov) [inline framework/db/QueryBuilder.php:620]: Can `$casesByColumn` be empty here?

File: framework/db/QueryBuilder.php [hunk 10]
```diff
@@ -501,6 +502,304 @@ public function batchInsert($table, $columns, $rows, &$params = [])
...
+        $seenKeyValues = [];
+        $rowKeyData = [];
+
+        foreach ($rows as $row) {
+            if ($row instanceof \Traversable) {
+                $row = iterator_to_array($row);
+            }
+            if (!is_array($row)) {
+                throw new InvalidArgumentException('Each batch update row must be an array.');
+            }
```
Comment (Tigrov) [inline framework/db/QueryBuilder.php]: This code can be called several times (in `normalizeBatchUpdateRows()` and in `resolveColumnNames()` if `$columns` not empty)
Perhaps it should be moved it in a separate method and called before `resolveColumnNames()`?

File: framework/db/QueryBuilder.php [hunk 11]
```diff
@@ -501,6 +502,304 @@ public function batchInsert($table, $columns, $rows, &$params = [])
...
+     *
+     * Override this method in DBMS-specific query builders to handle binary columns or other special types.
+     *
+     * @param string $table the table that data will be saved into.
+     * @param array $columns the column data (name => value) to be saved into the table.
+     * @param array $params the binding parameters that will be modified by this method.
+     * @return array normalized columns.
+     * @since 2.0.55
+     */
+    protected function normalizeTableRowData($table, $columns, &$params)
```
Comment (Tigrov) [inline framework/db/QueryBuilder.php]:
```suggestion
    protected function normalizeTableRowData($table, $columns)
```

Is it necessary to pass `$params` here? I haven't found where this is used.

File: framework/db/QueryBuilder.php [hunk 12]
```diff
@@ -501,6 +502,304 @@ public function batchInsert($table, $columns, $rows, &$params = [])
...
+     * (e.g. varbinary encoding of key values in MSSQL), then restored afterward.
+     *
+     * @param string $table the table name.
+     * @param array $rows the rows to normalize.
+     * @param array $keys the key column names.
+     * @param array $params the binding parameters that will be modified by this method.
+     * @return array normalized rows.
+     * @since 2.0.55
+     */
+    protected function normalizeBatchUpdateRows($table, $rows, $keys, &$params)
```
Comment (Tigrov) [inline framework/db/QueryBuilder.php]:
```suggestion
    protected function normalizeBatchUpdateRows($table, $rows, $keys)
```

Is it necessary to pass `$params` here? I haven't found where this is used.

File: framework/db/QueryBuilder.php (hunk 9 above)
Comment (WarLikeLaux) [reply framework/db/QueryBuilder.php:620]: Yes. If every row contains only key columns and no updatable columns, all rows are skipped by the `if (empty($row)) continue` check. `testBatchUpdateWithoutUpdatableColumns` covers this.

File: framework/db/QueryBuilder.php (hunk 8 above)
Comment (WarLikeLaux) [reply framework/db/QueryBuilder.php:567]: Technically possible for the SQL generation part (CASE/WHEN and WHERE). The problem is duplicate detection - keys are hashed via `var_export()` to catch duplicates, and Expression objects can't be reliably compared this way. Key columns are typically PK/unique where `dbTypecast()` returns scalars, so Expression keys are an edge case. Kept the exception for now.

File: framework/db/QueryBuilder.php (hunk 12 above)
Comment (WarLikeLaux) [reply framework/db/QueryBuilder.php]: You're right, neither mssql nor pgsql use `$params` in their overrides. Removed `&$params` from `normalizeTableRowData()` and `normalizeBatchUpdateRows()`.

File: framework/db/QueryBuilder.php (hunk 11 above)
Comment (WarLikeLaux) [reply framework/db/QueryBuilder.php]: You're right, neither mssql nor pgsql use `$params` in their overrides. Removed `&$params` from `normalizeTableRowData()` and `normalizeBatchUpdateRows()`.

File: framework/db/QueryBuilder.php (hunk 10 above)
Comment (WarLikeLaux) [reply framework/db/QueryBuilder.php]: Right. Extracted row preparation (Traversable-to-array + type check) into a separate step called once before `resolveColumnNames()`. Downstream methods now work with plain arrays.

**Review by Tigrov — APPROVED** [review]
👍

**Comment by WarLikeLaux** [comment]
@coderabbitai full review

File: framework/db/QueryBuilder.php [hunk 13]
```diff
@@ -501,6 +502,305 @@ public function batchInsert($table, $columns, $rows, &$params = [])
...
+
+        if ($condition !== '' && $condition !== []) {
+            $conditionSql = $this->buildCondition($condition, $params);
+            if ($conditionSql !== '') {
+                $whereSql = '(' . $whereSql . ') AND (' . $conditionSql . ')';
+            }
+        }
+
+        return $sql . ' WHERE ' . $whereSql;
+    }
```
[the remark this answers is not in the document]
Comment (WarLikeLaux) [reply framework/db/QueryBuilder.php:675]: @coderabbitai skipping - this is a linter complexity threshold, not a correctness issue, and the method's already approved and tested across drivers. The refactor risk outweighs the style win here.
